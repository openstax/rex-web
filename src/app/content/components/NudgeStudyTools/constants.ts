export const arrowDesktopHeight = 11.5;
export const arrowDesktopWidth = 15;
export const arrowMobileHeight = 6.2;
export const arrowMobileWidth = 5.2;
export const arrowRightMargin = 1;
export const arrowTopMargin = 1.6;

export const closeButtonSize = 4;
export const closeButtonDistanceFromContent = 1.6;
export const closeButtonMobileMargin = 3;

export const contentMarginTop = 1.6;
export const contentMarginLeft = 1;
export const contentWidth = 69;

export const mobileSpotlightPadding = .6;

export const cookieNudge = {
  // How many times nuge was shown
  counter: 'nudge_study_guides_counter',
  // When nudge was shown
  date: 'nudge_study_guides_date',
  // How many pages user has opened (opening the same page twice counts it as 2)
  pageCounter: 'nudge_study_guides_page_counter',
};

// Show nudge only for users which opened at least 2 pages
export const nudgeStudyToolsMinPageLimit = 2;

export const daysUntilCookieExpires = 365 * 20;

// Show nudge every 4 weeks: 4 * 7 days * 24 hours * 60 minutes * 60 seconds * 1000 ms
export const timeIntervalBetweenShowingNudgeInMs = 4 * 7 * 24 * 60 * 60 * 1000;
// Show nudge max 3 times
export const nudgeStudyToolsShowLimit = 3;
// Id of the element which should be highlighted
export const nudgeStudyToolsTargetId = 'nudge-study-tools';
export const mobileNudgeStudyToolsTargetId = 'mobile-menu-button';
