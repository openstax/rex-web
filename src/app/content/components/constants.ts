import theme from '../../theme';
import { BookWithOSWebData } from '../types';

export const bookBannerDesktopBigHeight = 13;
export const bookBannerDesktopMiniHeight = 7;
export const bookBannerMobileBigHeight = 10.4;
export const bookBannerMobileMiniHeight = 6;

export const scrollOffset = 3;

export const sidebarDesktopWidth = 37.5;
export const sidebarMobileWidth = 30.4;
export const sidebarTransitionTime = 300;

export const searchResultsBarDesktopWidth = 37.5;
export const searchResultsBarMobileWidth = 33.8;

export const toolbarIconColor = theme.color.primary.gray;

export const mobileSearchContainerMargin = 1;
export const toolbarWidth = 8;
export const toolbarSearchInputHeight = 3.2;
export const toolbarSearchInputMobileHeight = 3.2;
export const toolbarSearchInputMobileWidth = 18.5;
export const toolbarToggleSearchBarTextMobileWidth = 11.2;
export const toolbarButtonMargin = 2;
export const toolbarHrHeight = 0.1;
export const sidebarDesktopWithToolbarWidth = searchResultsBarDesktopWidth + toolbarWidth;
export const toolbarMobileSearchWrapperHeight = toolbarSearchInputMobileHeight
  + (mobileSearchContainerMargin * 2)
  + toolbarHrHeight;

export const topbarDesktopHeight = 5;
export const topbarMobileHeight = 4;
export const toolbarMobileExpandedHeight = topbarMobileHeight
  + toolbarMobileSearchWrapperHeight;

export const verticalNavbarMaxWidth = 8;

export const searchSidebarTopOffset = bookBannerMobileMiniHeight
  + toolbarMobileExpandedHeight;

export const contentTextWidth = 82.5;

export const mainContentBackground = '#fff';

export const maxContentGutter = 6;
export const contentWrapperMaxWidth = contentTextWidth + sidebarDesktopWidth + verticalNavbarMaxWidth;

export const defaultTheme = 'blue' as BookWithOSWebData['theme'];
