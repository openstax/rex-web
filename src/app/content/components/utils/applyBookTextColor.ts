import { css } from 'styled-components/macro';
import theme from '../../../theme';
import { BookWithOSWebData } from '../../types';

export const applyBookTextColor = (props: {colorSchema: BookWithOSWebData['theme'] | undefined }) =>
  props.colorSchema && css`
    color: ${theme.color.primary[props.colorSchema].foreground};
  `;
