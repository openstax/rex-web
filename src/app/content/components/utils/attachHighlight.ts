import Highlighter, { Highlight, SerializedHighlight } from '@openstax/highlighter';
import Sentry, { Severity } from '../../../../helpers/Sentry';

const attachHighlight = <T extends Highlight | SerializedHighlight>(
  highlight: T,
  highlighter: Highlighter,
  errorMsg?: (highlight: T) => string
): Highlight | undefined => {
  let attachedHighlight: Highlight | undefined;

  try {
    highlighter.highlight(highlight);
    attachedHighlight = highlighter.getHighlight(highlight.id);
  } catch (e) {
    Sentry.captureException(e);
  }

  if (!attachedHighlight || !attachedHighlight.isAttached()) {
    Sentry.captureException(
      new Error(errorMsg ?  errorMsg(highlight) : `Highlight with id: ${highlight.id} has not been attached.`),
      Severity.Warning
    );
    attachedHighlight = undefined;
  }

  return attachedHighlight;
};

export default attachHighlight;
