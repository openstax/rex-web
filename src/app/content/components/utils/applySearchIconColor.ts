import { css } from 'styled-components/macro';
import theme from '../../../theme';
import { toolbarIconColor } from '../constants';

const primaryColors = theme.color.primary;

export const applySearchIconColor = (colorSchema: keyof typeof primaryColors | null) =>
  colorSchema ? css`
    color: ${primaryColors[colorSchema].foreground};

    :hover,
    :focus {
        color: ${primaryColors[colorSchema].foreground};
        transition: color 200ms;

        ::before {
            content:"";
            display: block;
            height: 100%;
            position: absolute;
            top: 0;
            right: 0;
            min-width: 45px;
            background-color: rgba(0, 0, 0, 0.06)
            ;
        }
    }
  ` : css`
  color: ${toolbarIconColor.base};
`;
