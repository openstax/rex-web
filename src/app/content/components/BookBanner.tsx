import { HTMLAnchorElement, HTMLDivElement } from '@openstax/types/lib.dom';
import React from 'react';
import { useSelector } from 'react-redux';
import { FlattenSimpleInterpolation } from 'styled-components';
import styled, { css } from 'styled-components/macro';
import { ChevronLeft } from 'styled-icons/boxicons-regular/ChevronLeft';
import { maxNavWidth } from '../../components/NavBar';
import { h3MobileLineHeight, h3Style, h4Style, textRegularLineHeight } from '../../components/Typography';
import { useServices } from '../../context/Services';
import theme from '../../theme';
import { assertDefined, assertWindow } from '../../utils';
import { hasOSWebData } from '../guards';
import showConfirmation from '../highlights/components/utils/showConfirmation';
import { hasUnsavedHighlight as hasUnsavedHighlightSelector } from '../highlights/selectors';
import * as select from '../selectors';
import { BookWithOSWebData } from '../types';
import { isClickWithModifierKeys } from '../utils/domUtils';
import { bookDetailsUrl } from '../utils/urlUtils';
import {
  bookBannerDesktopBigHeight,
  bookBannerDesktopMiniHeight,
  bookBannerMobileBigHeight,
  bookBannerMobileMiniHeight,
  contentTextWidth
} from './constants';
import { applyBookTextColor } from './utils/applyBookTextColor';
import { disablePrint } from './utils/disablePrint';

const gradients: {[key in BookWithOSWebData['theme']]: string} = {
  'blue': '#004aa2',
  'deep-green': '#12A28C',
  'gray': '#97999b',
  'green': '#9cd14a',
  'light-blue': '#1EE1F0',
  'orange': '#FAA461',
  'red': '#E34361',
  'yellow': '#faea36',
};

// tslint:disable-next-line:variable-name
const LeftArrow = styled(ChevronLeft)`
  margin-top: -0.25rem;
  margin-left: -0.8rem;
  height: 3rem;
  width: 3rem;
  ${applyBookTextColor}
`;

// tslint:disable-next-line:variable-name
const TopBar = styled.div`
  width: 100%;
  max-width: ${maxNavWidth}rem;
  margin: 0 auto;
`;

const bookBannerTextStyle = css`
  max-width: ${maxNavWidth - (maxNavWidth - contentTextWidth) / 2}rem;
  padding: 0;
  ${applyBookTextColor}
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
`;

type Style = string | number | FlattenSimpleInterpolation;
const ifMiniNav = (miniStyle: Style, bigStyle?: Style) =>
  (props: {variant: 'mini' | 'big'}) =>
    props.variant === 'mini' ? miniStyle : bigStyle;

const bookTitleMiniNavDestkopWidth = 27;

const bookTitleStyles = css`
  ${h4Style}
  ${bookBannerTextStyle}
  display: ${ifMiniNav('inline-block', 'block')};
  height: ${textRegularLineHeight}rem;
  font-weight: normal;
  text-decoration: none;
  margin: 0;

  ${theme.breakpoints.mobile(css`
    ${bookBannerTextStyle}
  `)}

  ${ifMiniNav(css`
    width: ${bookTitleMiniNavDestkopWidth}rem;

    ${theme.breakpoints.mobile(css`
      display: none;
    `)}
  `)}
`;

// tslint:disable-next-line:variable-name
const BookTitle = styled.span`
  ${bookTitleStyles}
`;

// tslint:disable-next-line:variable-name
const BookTitleLink = styled.a`
  ${bookTitleStyles}
  :hover {
    text-decoration: underline;
  }
`;

// tslint:disable-next-line:variable-name
const BookChapter = styled(({colorSchema: _, variant, children, ...props}) => variant === 'mini' ?
  <span {...props}>{children}</span> : <h1 {...props}>{children}</h1>)`
  ${ifMiniNav(h4Style, h3Style)}
  ${bookBannerTextStyle}
  font-weight: 600;
  display: ${ifMiniNav('inline-block', 'block')};
  margin: 1rem 0 0 0;
  ${theme.breakpoints.mobile(css`
    ${bookBannerTextStyle}
    white-space: normal;
    display: -webkit-box;
    -webkit-box-orient: vertical;
    -webkit-line-clamp: 2;

    max-height: ${h3MobileLineHeight * 2}rem;
    margin-top: 0.3rem;
  `)}
  ${ifMiniNav(css`
    max-width: ${maxNavWidth - bookTitleMiniNavDestkopWidth - (maxNavWidth - contentTextWidth) / 2}rem;

    ${theme.breakpoints.mobile(css`
      max-width: none;
    `)}
  `)}
`;

interface BarWrapperProps {
  colorSchema: BookWithOSWebData['theme'] | undefined;
  up: boolean;
  variant: 'mini' | 'big';
}
// tslint:disable-next-line:variable-name
export const BarWrapper = styled.div<BarWrapperProps>`
  ${disablePrint}

  top: 0;
  padding: 0 ${theme.padding.page.desktop}rem;
  box-shadow: 0 0.2rem 0.2rem 0 rgba(0, 0, 0, 0.1);
  display: flex;
  align-items: center;
  height: ${ifMiniNav(bookBannerDesktopMiniHeight, bookBannerDesktopBigHeight)}rem;
  transition: transform 200ms;
  position: ${ifMiniNav('sticky', 'relative' /* stay above mini nav */)};
  z-index: ${ifMiniNav(theme.zIndex.navbar - 2, theme.zIndex.navbar - 1)};
  overflow: hidden;
  ${(props: {colorSchema: BookWithOSWebData['theme'] | undefined }) => props.colorSchema && css`
    background: linear-gradient(to right,
      ${assertDefined(
        theme.color.primary[props.colorSchema], `Could not find values for theme named "${props.colorSchema}"`
      ).base},
      ${assertDefined(
        gradients[props.colorSchema], `theme ${props.colorSchema} needs gradient defined in BookBanner.tsx`
      )});
  `}

  ${(props) => props.up && css`
    transform: translateY(-${bookBannerDesktopMiniHeight}rem);
    ${theme.breakpoints.mobile(css`
      transform: translateY(-${bookBannerMobileMiniHeight}rem);
    `)}
  `}

  ${theme.breakpoints.mobile(css`
    padding: ${theme.padding.page.mobile}rem;
    height: ${ifMiniNav(bookBannerMobileMiniHeight, bookBannerMobileBigHeight)}rem;
    ${ifMiniNav(`margin-top: -${bookBannerMobileMiniHeight}rem`)}
  `)}

  ${ifMiniNav(`margin-top: -${bookBannerDesktopMiniHeight}rem`)}
`;

interface BookBannerState {
  scrollTransition: boolean;
  tabbableBanner: 'mini' | 'big';
}

// tslint:disable-next-line:variable-name
const BookBanner = () => {
  const [bookBannerState, setBookBannerState] = React.useState<BookBannerState>({
    scrollTransition: false,
    tabbableBanner: 'big',
  });
  const book = useSelector(select.book);
  const treeSection = useSelector(select.pageNode);
  const bookTheme = useSelector(select.bookTheme);
  const hasUnsavedHighlight = useSelector(hasUnsavedHighlightSelector);
  const miniBanner = React.useRef<HTMLDivElement>();
  const bigBanner = React.useRef<HTMLDivElement>();
  const services = useServices();

  const handleScroll = () => {
    if (miniBanner.current && bigBanner.current && typeof(window) !== 'undefined') {
      const miniRect = miniBanner.current.getBoundingClientRect();

      setBookBannerState({
        scrollTransition: miniRect.top === 0
        && bigBanner.current.offsetTop + bigBanner.current.clientHeight > window.scrollY,
        tabbableBanner: miniRect.top === 0 ? 'mini' : 'big',
      });
    }
  };

  const handleLinkClick = async(e: React.MouseEvent<HTMLAnchorElement>, link: string) => {
    if (isClickWithModifierKeys(e) || !hasUnsavedHighlight) {
      return;
    }
    e.preventDefault();

    if (!await showConfirmation(services)) {
      return;
    }
    assertWindow().location.assign(link);
  };

  React.useEffect(() => {
    if (typeof document === 'undefined') {
      return;
    }
    document.addEventListener('scroll', handleScroll);
    handleScroll();

    return () => {
      document?.removeEventListener('scroll', handleScroll);
    };
  }, []);

  if (!book) {
    return <BarWrapper colorSchema={undefined} up={false} />;
  }

  const bookUrl = hasOSWebData(book)
    ? book.book_state !== 'retired'
      ? bookDetailsUrl(book)
      : undefined
    : undefined;

  return <>
    <BarWrapper
      colorSchema={bookTheme}
      key='expanded-nav'
      up={bookBannerState.scrollTransition}
      ref={bigBanner}
      data-testid='bookbanner'
      data-analytics-region='book-banner-expanded'
    >
      <TopBar>
        {
          bookUrl === undefined
            ? <BookTitle data-testid='book-title-expanded' colorSchema={bookTheme}>
              {book.tree.title}
            </BookTitle>
            : <BookTitleLink
              data-testid='details-link-expanded'
              href={bookUrl}
              colorSchema={bookTheme}
              onClick={(e: React.MouseEvent<HTMLAnchorElement>) => {
                handleLinkClick(e, bookUrl);
              }}
              tabIndex={bookBannerState.tabbableBanner === 'big' ? undefined : -1}
            >
              <LeftArrow colorSchema={bookTheme} />{book.tree.title}
            </BookTitleLink>
        }
        {treeSection
          ? <BookChapter
            colorSchema={bookTheme}
            dangerouslySetInnerHTML={{ __html: treeSection.title }}
          />
          : null}
      </TopBar>
    </BarWrapper>
    <BarWrapper
      colorSchema={bookTheme}
      variant='mini'
      key='mini-nav'
      ref={miniBanner}
      data-testid='bookbanner-collapsed'
      data-analytics-region='book-banner-collapsed'
    >
      <TopBar>
        {
          bookUrl === undefined
            ? <BookTitle data-testid='book-title-collapsed' colorSchema={bookTheme} variant='mini'>
              {book.title}
            </BookTitle>
            : <BookTitleLink
              data-testid='details-link-collapsed'
              href={bookUrl}
              variant='mini'
              colorSchema={bookTheme}
              onClick={(e: React.MouseEvent<HTMLAnchorElement>) => {
                handleLinkClick(e, bookUrl);
              }}
              tabIndex={bookBannerState.tabbableBanner === 'mini' ? undefined : -1}
            >
              <LeftArrow colorSchema={bookTheme} />{book.tree.title}
            </BookTitleLink>
        }
        {treeSection
          ? <BookChapter
            colorSchema={bookTheme}
            variant='mini'
            dangerouslySetInnerHTML={{ __html: treeSection.title }}
          />
          : null
        }
      </TopBar>
    </BarWrapper>
  </>;
};

export default BookBanner;
