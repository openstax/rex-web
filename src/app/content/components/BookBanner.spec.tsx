import { DOMRect } from '@openstax/types/lib.dom';
import { book as archiveBook, shortPage } from '../../../test/mocks/archiveLoader';
import { mockCmsBook } from '../../../test/mocks/osWebLoader';
import { renderToDom } from '../../../test/reactutils';
import { reactAndFriends, resetModules } from '../../../test/utils';
import { assertDocument, assertWindow } from '../../utils';
import { formatBookData } from '../utils';
import { findArchiveTreeNodeById } from '../utils/archiveTreeUtils';
import { BarWrapper, PropTypes } from './BookBanner';
import { defaultTheme } from './constants';

const book = formatBookData(archiveBook, mockCmsBook);
const bookWithoutOsWebData = formatBookData(archiveBook, undefined);
const pageNode = findArchiveTreeNodeById(archiveBook.tree, shortPage.id)!;

describe('BookBanner', () => {
  let window: Window;
  let event: React.MouseEvent;
  let assign: jest.SpyInstance;
  // tslint:disable-next-line:variable-name
  let BookBanner: React.ComponentType<PropTypes>;

  beforeEach(() => {
    resetModules();

    window = assertWindow();
    delete (window as any).location;

    window.location = {
      assign: jest.fn(),
    } as any as Window['location'];

    event = {
      altKey: false,
      ctrlKey: false,
      metaKey: false,
      preventDefault: jest.fn(),
      shiftKey: false,
    } as any as React.MouseEvent;

    assign = jest.spyOn(window.location, 'assign');
  });

  describe('without unsaved changes', () => {
    let React: any; // tslint:disable-line:variable-name
    let renderer: any;

    beforeEach(() => {
      ({React, renderer} = reactAndFriends());
      BookBanner = require('./BookBanner').BookBanner;
    });

    it('renders empty state with no page or book', () => {
      const component = renderer.create(<BookBanner bookTheme={defaultTheme} />);

      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('renders correctly when you pass a page and book', () => {
      const component = renderer.create(<BookBanner pageNode={pageNode} book={book} bookTheme={book.theme} />);

      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('renders BookTitle instead of BookTitleLink with a link to details page for retired books', () => {
      const component = renderer.create(<BookBanner
        pageNode={pageNode}
        book={{...book, book_state: 'retired'}}
        bookTheme={book.theme}
      />);

      expect(() => component.root.findByProps({ 'data-testid': 'book-title-expanded'})).not.toThrow();
      expect(() => component.root.findByProps({ 'data-testid': 'book-title-collapsed'})).not.toThrow();
      expect(() => component.root.findByProps({ 'data-testid': 'details-link-expanded'})).toThrow();
      expect(() => component.root.findByProps({ 'data-testid': 'details-link-collapsed'})).toThrow();
    });

    it('renders correctly without osweb data', () => {
      const component = renderer.create(<BookBanner
        pageNode={pageNode}
        book={bookWithoutOsWebData}
        bookTheme={defaultTheme}
      />);

      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('renders correctly without pageNode', () => {
      const component = renderer.create(<BookBanner book={book} bookTheme={defaultTheme} />);

      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('does not stop default navigation event', async() => {
      const component = renderer.create(<BookBanner
        pageNode={pageNode}
        book={book}
        bookTheme={book.theme}
        hasUnsavedHighlight={false}
      />);

      const link = component.root.findByProps({'data-testid': 'details-link-expanded'});

      await renderer.act(() => {
        link.props.onClick(event);
        return Promise.resolve();
      });

      expect(event.preventDefault).not.toHaveBeenCalled();
    });

    it('mounts in a dom', () => {
      expect(() => renderToDom(<BookBanner pageNode={pageNode} book={book} bookTheme={book.theme} />)).not.toThrow();
    });

    it('wrapper transition matches snapshot', () => {
      const component = renderer.create(<BarWrapper colorSchema='blue' up={true} />);
      expect(component.toJSON()).toMatchSnapshot();
    });

    it('defaults tab indexes on banner links', () => {
      const component = renderer.create(<BookBanner pageNode={pageNode} book={book} bookTheme={book.theme} />);

      const linkExpanded = component.root.findByProps({'data-testid': 'details-link-expanded'});
      const linkCollapsed = component.root.findByProps({'data-testid': 'details-link-collapsed'});

      expect(linkExpanded.props.tabIndex).toBe(undefined);
      expect(linkCollapsed.props.tabIndex).toBe(-1);
    });

    it('sets tab indexes on banner links according to scroll', () => {
      const expandedBannerNode = assertDocument().createElement('div');
      const collapsedBannerNode = assertDocument().createElement('div');

      const createNodeMock = (element: any) => {
        const analyticsRegion = element.props['data-testid'];

        if (analyticsRegion === 'bookbanner') {
          return expandedBannerNode;
        }
        if (analyticsRegion === 'bookbanner-collapsed') {
          return collapsedBannerNode;
        }

        return null;
      };

      // scroll handler also gets called on component did mount
      const rectSpy = jest.spyOn(collapsedBannerNode, 'getBoundingClientRect')
        .mockReturnValueOnce({top: 50} as DOMRect);

      const component = renderer.create(
        <BookBanner pageNode={pageNode} book={book} bookTheme={book.theme}/>,
        {createNodeMock}
      );

      const linkExpanded = component.root.findByProps({'data-testid': 'details-link-expanded'});
      const linkCollapsed = component.root.findByProps({'data-testid': 'details-link-collapsed'});

      const scrollEvent = window.document.createEvent('UIEvents');
      scrollEvent.initEvent('scroll', true, false);

      // first scroll
      rectSpy.mockReturnValueOnce({top: 0} as DOMRect);
      renderer.act(() => {
        window.document.dispatchEvent(scrollEvent);
      });

      expect(linkExpanded.props.tabIndex).toBe(-1);
      expect(linkCollapsed.props.tabIndex).toBe(undefined);

      // second scroll
      rectSpy.mockReturnValueOnce({top: 50} as DOMRect);
      renderer.act(() => {
        window.document.dispatchEvent(scrollEvent);
      });

      expect(linkExpanded.props.tabIndex).toBe(undefined);
      expect(linkCollapsed.props.tabIndex).toBe(-1);
    });
  });

  describe('with unsaved changes', () => {
    const mockConfirmation = jest.fn()
      .mockImplementationOnce(() =>  new Promise((resolve) => resolve(true)))
      .mockImplementationOnce(() =>  new Promise((resolve) => resolve(false)));

    jest.mock(
      '../highlights/components/utils/showConfirmation',
      () => mockConfirmation
    );

    let React: any; // tslint:disable-line:variable-name
    let renderer: any;

    beforeEach(() => {
      ({React, renderer} = reactAndFriends());
      BookBanner = require('./BookBanner').BookBanner;
    });

    it('redirects if users chooses to discard', async() => {
      const component = renderer.create(<BookBanner
        pageNode={pageNode}
        book={book}
        bookTheme={book.theme}
        hasUnsavedHighlight={true}
      />);

      const link = component.root.findByProps({'data-testid': 'details-link-expanded'});

      await renderer.act(() => {
        link.props.onClick(event);
        return Promise.resolve();
      });

      expect(event.preventDefault).toHaveBeenCalled();
      expect(assign).toHaveBeenCalled();
    });

    it('noops if users chooses not to discard', async() => {
      const component = renderer.create(<BookBanner
        pageNode={pageNode}
        book={book}
        bookTheme={book.theme}
        hasUnsavedHighlight={true}
      />);

      const link = component.root.findByProps({'data-testid': 'details-link-collapsed'});

      await renderer.act(() => {
        link.props.onClick(event);
        return Promise.resolve();
      });

      expect(event.preventDefault).toHaveBeenCalled();
      expect(assign).not.toHaveBeenCalled();
    });
  });

  describe('outside browser', () => {
    const windowBackup = window;
    const documentBackup = document;
    let React: any; // tslint:disable-line:variable-name
    let renderer: any;

    beforeEach(() => {
      delete (global as any).window;
      delete (global as any).document;
      ({React, renderer} = reactAndFriends());
    });

    afterEach(() => {
      (global as any).window = windowBackup;
      (global as any).document = documentBackup;
    });

    it('renders', () => {
      const component = renderer.create(<BookBanner pageNode={pageNode} book={book} bookTheme={book.theme} />);

      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
