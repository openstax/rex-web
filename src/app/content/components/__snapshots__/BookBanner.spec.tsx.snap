// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookBanner outside browser renders 1`] = `
Array [
  <div
    className="BookBanner__BarWrapper-sc-1avy0c0-5 MNbyW"
    data-analytics-region="book-banner-expanded"
    data-testid="bookbanner"
  >
    <div
      className="BookBanner__TopBar-sc-1avy0c0-1 jVswHX"
    >
      <a
        className="BookBanner__BookTitleLink-sc-1avy0c0-3 iGPDns"
        data-testid="details-link-expanded"
        href="/details/books/book-slug-1"
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          className="BookBanner__LeftArrow-sc-1avy0c0-0 jjTVXA StyledIconBase-sc-bdy9j4 jbPeIR"
          fill="currentColor"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M13.293 6.293L7.586 12l5.707 5.707 1.414-1.414L10.414 12l4.293-4.293z"
          />
        </svg>
        Test Book 1
      </a>
      <h1
        className="BookBanner__BookChapter-sc-1avy0c0-4 kPUGut"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<span class=\\"os-text\\">Test Page 4</span>",
          }
        }
      />
    </div>
  </div>,
  <div
    className="BookBanner__BarWrapper-sc-1avy0c0-5 bheBBm"
    data-analytics-region="book-banner-collapsed"
    data-testid="bookbanner-collapsed"
  >
    <div
      className="BookBanner__TopBar-sc-1avy0c0-1 jVswHX"
    >
      <a
        className="BookBanner__BookTitleLink-sc-1avy0c0-3 iJDVtP"
        data-testid="details-link-collapsed"
        href="/details/books/book-slug-1"
        onClick={[Function]}
        tabIndex={-1}
      >
        <svg
          aria-hidden="true"
          className="BookBanner__LeftArrow-sc-1avy0c0-0 jjTVXA StyledIconBase-sc-bdy9j4 jbPeIR"
          fill="currentColor"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M13.293 6.293L7.586 12l5.707 5.707 1.414-1.414L10.414 12l4.293-4.293z"
          />
        </svg>
        Test Book 1
      </a>
      <span
        className="BookBanner__BookChapter-sc-1avy0c0-4 fNXDFo"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<span class=\\"os-text\\">Test Page 4</span>",
          }
        }
      />
    </div>
  </div>,
]
`;

exports[`BookBanner without unsaved changes renders correctly when you pass a page and book 1`] = `
Array [
  <div
    className="BookBanner__BarWrapper-sc-1avy0c0-5 MNbyW"
    data-analytics-region="book-banner-expanded"
    data-testid="bookbanner"
  >
    <div
      className="BookBanner__TopBar-sc-1avy0c0-1 jVswHX"
    >
      <a
        className="BookBanner__BookTitleLink-sc-1avy0c0-3 iGPDns"
        data-testid="details-link-expanded"
        href="/details/books/book-slug-1"
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          className="BookBanner__LeftArrow-sc-1avy0c0-0 jjTVXA StyledIconBase-sc-bdy9j4 jbPeIR"
          fill="currentColor"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M13.293 6.293L7.586 12l5.707 5.707 1.414-1.414L10.414 12l4.293-4.293z"
          />
        </svg>
        Test Book 1
      </a>
      <h1
        className="BookBanner__BookChapter-sc-1avy0c0-4 kPUGut"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<span class=\\"os-text\\">Test Page 4</span>",
          }
        }
      />
    </div>
  </div>,
  <div
    className="BookBanner__BarWrapper-sc-1avy0c0-5 bheBBm"
    data-analytics-region="book-banner-collapsed"
    data-testid="bookbanner-collapsed"
  >
    <div
      className="BookBanner__TopBar-sc-1avy0c0-1 jVswHX"
    >
      <a
        className="BookBanner__BookTitleLink-sc-1avy0c0-3 iJDVtP"
        data-testid="details-link-collapsed"
        href="/details/books/book-slug-1"
        onClick={[Function]}
        tabIndex={-1}
      >
        <svg
          aria-hidden="true"
          className="BookBanner__LeftArrow-sc-1avy0c0-0 jjTVXA StyledIconBase-sc-bdy9j4 jbPeIR"
          fill="currentColor"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M13.293 6.293L7.586 12l5.707 5.707 1.414-1.414L10.414 12l4.293-4.293z"
          />
        </svg>
        Test Book 1
      </a>
      <span
        className="BookBanner__BookChapter-sc-1avy0c0-4 fNXDFo"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<span class=\\"os-text\\">Test Page 4</span>",
          }
        }
      />
    </div>
  </div>,
]
`;

exports[`BookBanner without unsaved changes renders correctly without osweb data 1`] = `
Array [
  <div
    className="BookBanner__BarWrapper-sc-1avy0c0-5 MNbyW"
    data-analytics-region="book-banner-expanded"
    data-testid="bookbanner"
  >
    <div
      className="BookBanner__TopBar-sc-1avy0c0-1 jVswHX"
    >
      <span
        className="BookBanner__BookTitle-sc-1avy0c0-2 gUeKV"
        data-testid="book-title-expanded"
      >
        Test Book 1
      </span>
      <h1
        className="BookBanner__BookChapter-sc-1avy0c0-4 kPUGut"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<span class=\\"os-text\\">Test Page 4</span>",
          }
        }
      />
    </div>
  </div>,
  <div
    className="BookBanner__BarWrapper-sc-1avy0c0-5 bheBBm"
    data-analytics-region="book-banner-collapsed"
    data-testid="bookbanner-collapsed"
  >
    <div
      className="BookBanner__TopBar-sc-1avy0c0-1 jVswHX"
    >
      <span
        className="BookBanner__BookTitle-sc-1avy0c0-2 jvAwFk"
        data-testid="book-title-collapsed"
      >
        Test Book 1
      </span>
      <span
        className="BookBanner__BookChapter-sc-1avy0c0-4 fNXDFo"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<span class=\\"os-text\\">Test Page 4</span>",
          }
        }
      />
    </div>
  </div>,
]
`;

exports[`BookBanner without unsaved changes renders correctly without pageNode 1`] = `
Array [
  <div
    className="BookBanner__BarWrapper-sc-1avy0c0-5 MNbyW"
    data-analytics-region="book-banner-expanded"
    data-testid="bookbanner"
  >
    <div
      className="BookBanner__TopBar-sc-1avy0c0-1 jVswHX"
    >
      <a
        className="BookBanner__BookTitleLink-sc-1avy0c0-3 iGPDns"
        data-testid="details-link-expanded"
        href="/details/books/book-slug-1"
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          className="BookBanner__LeftArrow-sc-1avy0c0-0 jjTVXA StyledIconBase-sc-bdy9j4 jbPeIR"
          fill="currentColor"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M13.293 6.293L7.586 12l5.707 5.707 1.414-1.414L10.414 12l4.293-4.293z"
          />
        </svg>
        Test Book 1
      </a>
    </div>
  </div>,
  <div
    className="BookBanner__BarWrapper-sc-1avy0c0-5 bheBBm"
    data-analytics-region="book-banner-collapsed"
    data-testid="bookbanner-collapsed"
  >
    <div
      className="BookBanner__TopBar-sc-1avy0c0-1 jVswHX"
    >
      <a
        className="BookBanner__BookTitleLink-sc-1avy0c0-3 iJDVtP"
        data-testid="details-link-collapsed"
        href="/details/books/book-slug-1"
        onClick={[Function]}
        tabIndex={-1}
      >
        <svg
          aria-hidden="true"
          className="BookBanner__LeftArrow-sc-1avy0c0-0 jjTVXA StyledIconBase-sc-bdy9j4 jbPeIR"
          fill="currentColor"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M13.293 6.293L7.586 12l5.707 5.707 1.414-1.414L10.414 12l4.293-4.293z"
          />
        </svg>
        Test Book 1
      </a>
    </div>
  </div>,
]
`;

exports[`BookBanner without unsaved changes renders empty state with no page or book 1`] = `
<div
  className="BookBanner__BarWrapper-sc-1avy0c0-5 blPrsP"
/>
`;

exports[`BookBanner without unsaved changes wrapper transition matches snapshot 1`] = `
<div
  className="BookBanner__BarWrapper-sc-1avy0c0-5 bYEYED"
/>
`;
